src/crypto.rs

```diff
@@ -72,6 +122,349 @@ pub fn verify_signed_xml<Bytes: AsRef<[u8]>>(
+                    let obj_ptr = unsafe {
+                        libxml::bindings::xmlXPtrEval(
+                            c_uri.as_ptr() as *const libxml::bindings::xmlChar,
+                            ctx.pointer,
+                        )
+                    };
+                    if obj_ptr.is_null() {
+                        return None;
+                    }
+                    let obj = XPathObject { pointer: obj_ptr };
+
+                    // extract the nodeset from the result
+                    let obj_type = unsafe { (*obj.pointer).type_ };
+                    if obj_type != libxml::bindings::xmlXPathObjectType_XPATH_NODESET {
+                        return None;
+                    }
+                    let obj_nodeset = unsafe { (*obj.pointer).nodesetval };
+                    let nodeset_count = unsafe { (*obj_nodeset).nodeNr };
+
+                    // go through the nodes and find them in the document
+                    for i in 0..nodeset_count {
+                        let node_ptr_ptr = unsafe { (*obj_nodeset).nodeTab.offset(i.try_into().unwrap()) };
+                        let node_ptr = unsafe { *node_ptr_ptr };
+                        if let Some(node) = get_node_by_ptr(&doc.get_root_element().unwrap(), node_ptr) {
+                            return Some(node);
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    None
+}
+
+/// Place the signature-verified attributes ([`ATTRIB_SIGVER`] in the given namespace) on the given
+/// element, all its descendants and its whole chain of ancestors (but not necessarily all their
+/// descendants).
+#[cfg(feature = "xmlsec")]
+fn place_signature_verified_attributes(
+    root_elem: libxml::tree::Node,
+    doc: &libxml::tree::Document,
+    ns: &libxml::tree::Namespace,
+) {
+    let mut ptr_to_required_node: HashMap<usize, libxml::tree::Node> = HashMap::new();
+    let mut signature_nodes = find_signature_nodes(&root_elem);
+    for sig_node in signature_nodes.drain(..) {
+        if let Some(sig_root_node) = get_signed_node(&sig_node, doc) {
+            let mut nodes = Vec::new();
+            let mut parent = sig_root_node.get_parent();
+            nodes.push(sig_root_node);
+
+            // mark all children
+            while nodes.len() > 0 {
+                let node = nodes.pop().unwrap();
+                let node_ptr = node.node_ptr() as usize;
+                for child in node.get_child_elements() {
+                    nodes.push(child);
+                }
+                ptr_to_required_node.entry(node_ptr).or_insert(node);
+            }
+
+            // mark the ancestor chain
+            while let Some(p) = parent {
+                let p_ptr = p.node_ptr() as usize;
+                parent = p.get_parent();
+                ptr_to_required_node.entry(p_ptr).or_insert(p);
+            }
+        }
+    }
+    drop(root_elem);
```

Comment:
I'm pretty unhappy with it as well, but `rust-libxml` forbids modifying a node if too many references to it are floating around -- basically a customized runtime version of the borrow checker -- so I'm forced to drop them at strategic points.

In this case, since one of the nodes in `ptr_to_required_node` in the loop right after this `drop` call might be the root node and modifying it will fail if I still hold the reference in `root_elem`, I have to get rid of it first...

Reply:
Dropping the last reference to the root element will (unsafely) recursively drop all child elements, so without some kind of correctness proof, I don't think we should be dropping it here.

Can we just not add the root elem to the map, and instead of a `bool` tracking whether the root needs this to happen as well?

Reply:
As I understand it, all nodes are still referenced within `Document` and therefore immune to disappearing completely.

I tried to turn the `root_elem` argument of `place_signature_verified_attributes` into a mutable reference; this raised my favorite error with too many references. I tried the workaround with the `bool` value (ensuring that all other values except the arguments went out of scope beforehand) and got the same error. Sorry. ☹️